lib/grammers-session/src/message_box/mod.rs

```diff
@@ -568,59 +601,88 @@ impl MessageBox {
                 debug!(
                     "handling too-long difference (pts = {}); no longer getting diff",
                     diff.pts
                 );
-                // TODO when are deadlines reset if we update the map??
+                finish = true;
+                // the deadline will be reset once the diff ends
                 self.map.get_mut(&Entry::AccountWide).unwrap().pts = diff.pts;
+                result = (Vec::new(), Vec::new(), Vec::new())
+            }
+        }
+
+        if finish {
+            let account = self.getting_diff_for.get(&Entry::AccountWide).is_some();
+            let secret = self.getting_diff_for.get(&Entry::SecretChats).is_some();
+
+            if !account && !secret {
+                panic!("Should not be applying the difference when neither account or secret diff was active")
+            }
+
+            if account {
                 self.end_get_diff(Entry::AccountWide);
-                (Vec::new(), Vec::new(), Vec::new())
+            }
+            if secret {
+                self.end_get_diff(Entry::SecretChats);
             }
         }
+
+        Ok(result)
     }
 
     fn apply_difference_type(
         &mut self,
         tl::types::updates::Difference {
             new_messages,
             new_encrypted_messages,
-            other_updates: mut updates,
+            other_updates: updates,
             chats,
             users,
             state: tl::enums::updates::State::State(state),
         }: tl::types::updates::Difference,
-    ) -> (
-        Vec<tl::enums::Update>,
-        Vec<tl::enums::User>,
-        Vec<tl::enums::Chat>,
-    ) {
+        chat_hashes: &mut ChatHashCache,
+    ) -> Result<
+        (
+            Vec<tl::enums::Update>,
+            Vec<tl::enums::User>,
+            Vec<tl::enums::Chat>,
+        ),
+        Gap,
+    > {
         self.map.get_mut(&Entry::AccountWide).unwrap().pts = state.pts;
         self.map.get_mut(&Entry::SecretChats).unwrap().pts = state.qts;
         self.date = state.date;
         self.seq = state.seq;
 
-        updates.iter().for_each(|u| match u {
-            tl::enums::Update::ChannelTooLong(c) => {
-                // `c.pts`, if any, is the channel's current `pts`; we do not need this.
-                info!("got {:?} during getDifference", c);
-                self.begin_get_diff(Entry::Channel(c.channel_id));
-            }
-            _ => {}
+        // other_updates can contain things like UpdateChannelTooLong and UpdateNewChannelMessage.
+        // We need to process those as if they were socket updates to discard any we have already handled.
+        let mut result_updates = vec![];
+        let us = tl::enums::Updates::Updates(tl::types::Updates {
+            updates,
+            users: users.clone(),
+            chats: chats.clone(),
+            date: 1,
+            seq: NO_SEQ,
         });
+        self.process_updates(us, chat_hashes, &mut result_updates)?;
```

Comment:
In theory, applying the updates from `getDifference` should **not** trigger another gap (that's the very thing `getDifference` is supposed to be used for!).

I see that the client was updated to log an error if for some reason this does happen. Personally I'd be more inclined to panic here (the server must be misbehaving if it's returning gaps from `getDifference`), but I guess we can leave it to log an error for now and we'll see if it ever happens.

I'm also not sure how the client should behave if this does happen. I'm inclined to panic because we're probably not capable of handling `getDifference` returning gaps.

Reply:
Done